### PR TITLE
feat(bitmap): Move bitmap (port allocation tracking) to common libs. …

### PR DESCRIPTION
**Pull request by james-munson**
… Extend to avoid immediate reuse of freed.

#### Which issue(s) this PR fixes:
Part 1 of https://github.com/longhorn/longhorn/issues/8598.
Part 2 would be to remove the duplicated bitmap code from other repos and point them at this one.

#### Special notes for your reviewer:

For a handy diff of the new code against the old instance-manager version, see https://github.com/longhorn/longhorn-instance-manager/pull/625/files#diff-c65bce8b6b1aef67ab3a2a6dcd87b0834084a68d06b9f1ce1f127ab8612df22a

### CODE REVIEW

File: bitmap/bitmap.go
Comment (ejweber) [inline bitmap/bitmap.go:86]: This is the one line I struggle to understand, though it looks like it works fine in the tests. Is the idea that `b.start` may be too big now (`rEnd` may be the last value in the map), but it's fine, because we'll just fail out of the first AllocateRange next time and potentially find values in the fallback?

File: bitmap/bitmap.go
Comment (ejweber) [inline bitmap/bitmap.go:82]: NIT:
```suggestion
		return 0, 0, fmt.Errorf("cannot find an empty port range")
```

File: bitmap/bitmap.go
Comment (ejweber) [inline bitmap/bitmap.go:86]: This is the one line I struggle to understand, though it looks like it works fine in the tests. Is the idea that `b.start` may be too big now (`rEnd` may be the last value in the map), but it's fine, because we'll just fail out of the first `AllocateRange` next time and potentially find values in the fallback?

**Review by ejweber — COMMENTED** [review]
LGTM in principle.
- One question that I think is only for my own understanding.
- One nit for our standard error formatting conventions.

It looks like we have pretty strong checking for code coverage in this library compared to some of our other repos! The Codecov warnings are pretty pedantic here, but they seem like they can easily be covered by a few additional test cases with "bad" inputs. WDYT about handling them before merging?

File: bitmap/bitmap.go
Comment (james-munson) [reply bitmap/bitmap.go:86]: Yes, I should remove that comment.  I thought it might be necessary early on, but as you note, the restart from 0 takes care of that.  I just failed to clean it up.

**Comment by james-munson** [comment]
Thanks for the review.  The gaps in coverage were there when I got there, but as you say, it is the work of a moment to cure that as well, so why not?

**Review by ejweber — APPROVED** [review]
LGTM.

**Comment by ejweber** [comment]
The merge is blocked here for a similar reason to https://github.com/longhorn/longhorn-ui/pull/679#issuecomment-1901574306. It looks like this repo is also configured to reguire GPG (or other) signing? I don't have access to the settings to check the details. Is this expected @derekbit / @innobead?

(BTW it is fine to wait for @PhanLe1010 or @shuo-wu to take a look as well before we merge. Eventually though, we will need to unblock this.)

**Comment by james-munson** [comment]
Confusing.  The commit shows `Signed-off-by: James Munson <james.munson@suse.com>`.  I ensured it was there when doing the squash and re-push.

File: bitmap/bitmap.go
Comment (shuo-wu) [inline bitmap/bitmap.go:61]: I just realized that the partially allocated ports above cannot be released if the function errors out here.
This issue exists in the previous implementation as well.

File: bitmap/bitmap.go
Comment (shuo-wu) [inline bitmap/bitmap.go:36]: What if `iStart` is the largest number in `b.data`

Comment (james-munson) [reply bitmap/bitmap.go:36]: It's harmless. There is no panic in AdvanceIfNeeded().  HasNext() will fail, no range will be found, and the outer call will start again from 0.  The unit tests actually test this, having allocated up to the top, and then trying one more.

**Comment by derekbit** [comment]
> The merge is blocked here for a similar reason to [longhorn/longhorn-ui#679 (comment)](https://github.com/longhorn/longhorn-ui/pull/679#issuecomment-1901574306). It looks like this repo is also configured to reguire GPG (or other) signing? I don't have access to the settings to check the details. Is this expected @derekbit / @innobead?
> 
> (BTW it is fine to wait for @PhanLe1010 or @shuo-wu to take a look as well before we merge. Eventually though, we will need to unblock this.)

> It looks like this repo is also configured to reguire GPG (or other) signing?

No, we don't need it and configure it.

Unfortunately, I'm encountering the same issue. Figuring out the root cause...
image

cc @innobead

File: bitmap/bitmap.go
Comment (derekbit) [inline bitmap/bitmap.go:20]: How about adding input checks for `start` and `end`?

File: bitmap/bitmap.go
Comment (james-munson) [reply bitmap/bitmap.go:80]: That's right.  This is the whole point of the change.

File: bitmap/bitmap.go
Comment (james-munson) [reply bitmap/bitmap.go:20]: And do what?  Log it?  Return an error?  I was careful to keep the external interface the same as the original.  Callers such as instance-manager and backing image manager call NewBitmap in their own constructors, having first done the necessary validation with a method such as `ParsePortRange`.

Since this is a common library, NewBitmap can return (*Bitmap, error) and I can then rearrange the places it is called.  WDYT?

File: bitmap/bitmap.go
Comment (james-munson) [reply bitmap/bitmap.go:24]: Not as used today, but that could be enforced.  See previous comment.

File: bitmap/bitmap.go
Comment (james-munson) [reply bitmap/bitmap.go:106]: Done.

File: bitmap/bitmap.go
Comment (james-munson) [reply bitmap/bitmap.go]: Done.

File: bitmap/bitmap.go
Comment (james-munson) [reply bitmap/bitmap.go:19]: The bitmap is a range of values, not indexes.  Callers will create it with a port range such as (10000, 20000).

File: bitmap/bitmap.go
Comment (derekbit) [reply bitmap/bitmap.go:20]: Ideally, the function should return an error if the parameters are invalid. However, since go-common-libs will be used by Longhorn components across different versions, adding an `error` in returned values breaks compatibility. Therefore, the only option is to return nil and log the error. WDYT?

File: bitmap/bitmap.go
Comment (derekbit) [reply bitmap/bitmap.go:19]: Got it. Thanks.

File: bitmap/bitmap.go
Comment (derekbit) [reply bitmap/bitmap.go:80]: Could you add a comment here?

File: bitmap/bitmap.go
Comment (james-munson) [reply bitmap/bitmap.go:24]: Done.

File: bitmap/bitmap.go
Comment (james-munson) [reply bitmap/bitmap.go:85]: Done.

File: bitmap/bitmap_test.go
Comment (shuo-wu) [inline bitmap/bitmap_test.go:34]: For `bm, err = NewBitmap(100, 200)`, is port `200` included?

**Review by shuo-wu — APPROVED** [review]
In general LGTM

File: bitmap/bitmap_test.go
Comment (james-munson) [reply bitmap/bitmap_test.go:34]: Yes, it is.

**Review by derekbit — APPROVED** [review]

**Pull request merged**
